### PR TITLE
[Core] Connectivity preser modeler - bugfix - memory error

### PR DESCRIPTION
**Pull request by ddiezrod**
**📝 Description**
This was found some time ago by @pooyan-dadvand and @pablobecker , but it was never pushed here. The problem is that we are pointing to the same PropertiesContainer after we call the ConnectivityPreserveModeler. If one of the properties is changed or replaced in one of the model parts it will also change in the other model part.

**🆕 Changelog**
- Bugfix when assigning the properties to the destination model part.

### CODE REVIEW

**Comment by ddiezrod** [comment]
There is a test failing in ConvectionDiffusionApp exemplifies perfectly what the problem is. This is what was actually happening in this thermo-mechanical test.
- Thermal model part is generated from the structural model part using connectivity preserver modeler. Therefore, they point to the same properties container.
- Structural solver reads its StructuralMaterials.json adding the property 1.
- As both model parts share their properties container, thermal model part also has this new property at this point.
-  Thermal solver reads its ConvectionDiffusionMaterials.json which also has property 1, but with a different DENSITY value. ReadMaterialsUtility finds that property 1 already exists so it throws a warning saying that this property will be replaced.
- Again, as the structure model part and the thermal model part have the same properties container, the properties of the structural model part are actually changed now, so it's using the density from ConvectionDiffusionMaterials.json.

With the current change, thermal model part will have pointers to all the properties that were in the structural model part at the moment it is originated, but the container itself is different, so when new properties are added or replaced in any of the model parts, it will not change the other one.

**Comment by rubenzorrilla** [comment]
But isn't this the expected behavior? I mean, don't we want to have the same properties container in both model parts e.g. you have just one `DENSITY` value, not one density for the mechanical problem and a different one for the thermal problem.

**Review by RiccardoRossi — COMMENTED** [review]
uff this is a subtle one. I see both points of view, and to be honest i am not sure of the correct behaviour

   @KratosMultiphysics/technical-committee ?

**Comment by ddiezrod** [comment]
> But isn't this the expected behavior? I mean, don't we want to have the same properties container in both model parts e.g. you have just one `DENSITY` value, not one density for the mechanical problem and a different one for the thermal problem.

I think this is quite counterintuitive when the whole structure here is designed so each solver has its own model part that reads its own material properties json. Moreover, maybe in the case of the density is ok to have the same value in all model parts but I think it would not be so difficult to find cases where you want to have different values (e.g. the constitutive law (!!), (if Im not missing something))

**Comment by rubenzorrilla** [comment]
> > But isn't this the expected behavior? I mean, don't we want to have the same properties container in both model parts e.g. you have just one `DENSITY` value, not one density for the mechanical problem and a different one for the thermal problem.
> 
> I think this is quite counterintuitive when the whole structure here is designed so each solver has its own model part that reads its own material properties json. Moreover, maybe in the case of the density is ok to have the same value in all model parts but I think it would not be so difficult to find cases where you want to have different values (e.g. the constitutive law (!!), (if Im not missing something))

I see your point and it makes perfect sense (`DENSITY` and `ConstitutiveLaw` pointer are two very good counter examples). Having said this, I support @RiccardoRossi 's observation, there is no "correct" behavior at all but we need to choose one. As this is sensitive, I think we should briefly discuss it in the KTC (adding to the next meeting TODO list).

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee discussed about this change and the conclusion is to add a setting parameter which tells if copy the properties or assign the same one. 

The default value must keep the current behavior.

@KratosMultiphysics/implementation-committee would you please take care of the implementation of it?

**Comment by pooyan-dadvand** [comment]
Talking with @ddiezrod the problem is that we are sharing the container of all properties between modelparts not the properties themselves. This make it very vulnerable to any change afterward by adding or removing a properties to/from it.

For example if another solver wants to change the properties 1 with a new one with given CL. Then it is changing the one of the original one also. So then in the original modelpart you will end up with some properties that the elements are not pointing to.

File: kratos/modeler/connectivity_preserve_modeler.cpp [hunk 1]
```diff
@@ -144,7 +144,11 @@ void ConnectivityPreserveModeler::CopyCommonData(
 
     // These should be safe for SubModelParts
     rDestinationModelPart.SetProcessInfo( rOriginModelPart.pGetProcessInfo() );
-    rDestinationModelPart.SetProperties( rOriginModelPart.pProperties() );
```
Comment (RiccardoRossi) [inline kratos/modeler/connectivity_preserve_modeler.cpp:147]: @ddiezrod just to be sure we @KratosMultiphysics/technical-committee  understand the change you propose:

before we were ensuring that the container of the properties was the same between the original 
model_part and the new one.






after this change you are telling that the container are not the same, however the new container is populated with pointers to the same properties. Is that so?

-- OLD BEHAVIOUR --
origin_mp.Properties[1][DENSITY]=1

new_mp.Properties[1] has density 1

if i add a new property to either the new or the old, the new property will be seen identical in the two places.

-- NEW PROPOSED BEHAVIOUR --
pointers are initally the same, so if one changes density in one, also the other can feel it.
however, we could add a new property to one of the modelparts (or possibly reassign a property) to one of the two WITHOUT AFFECTING the other one

is that correct?

Comment (ddiezrod) [reply kratos/modeler/connectivity_preserve_modeler.cpp:147]: that is correct! @miguelmaso opened a MR that is equivalent to this one, adding a test that explains the new behavior. We can close this and continue the discussion there as I think that one is cleaner.

Comment (ddiezrod) [reply kratos/modeler/connectivity_preserve_modeler.cpp:147]: https://github.com/KratosMultiphysics/Kratos/pull/9758

**Comment by ddiezrod** [comment]
closing this as it is repeated in https://github.com/KratosMultiphysics/Kratos/pull/9758

**Comment by loumalouomega** [comment]
Do we still need this branch @ddiezrod ?, can we remove it?

**Pull request closed without merging**